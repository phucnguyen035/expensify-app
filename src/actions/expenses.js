import uuidv4 from 'uuid/v4';
import db from '../database/firebase';

// ADD_EXPENSE
const addExpense = newExpense => ({
  type: 'ADD_EXPENSE',
  expense: newExpense
});

const startAddExpense = (expenseData = {}) => (dispatch) => {
  const { amount, createdAt, desc, note } = expenseData;
  const newExpense = { amount, createdAt, desc, note };

  db.ref('expenses')
    .push(newExpense)
    .then((ref) => {
      dispatch(
        addExpense({
          id: ref.key,
          ...newExpense
        })
      );
    });
};

// REMOVE_EXPENSE
const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export { addExpense, startAddExpense, removeExpense, editExpense };
