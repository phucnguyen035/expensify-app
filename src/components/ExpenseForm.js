import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      desc: props.desc ? props.desc : '',
      note: props.note ? props.note : '',
      amount: props.amount ? (props.amount / 100).toString() : '',
      createdAt: props.createdAt ? moment(props.createdAt) : moment(),
      focused: false,
      error: ''
    };
  }

  onDescChange = (e) => {
    const desc = e.target.value;

    this.setState(() => ({ desc }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;

    this.setState(() => ({ note }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    // Match amount with number only with 2 decimal places
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ focused }));
  };

  onFormSubmit = (e) => {
    e.preventDefault();
    if (!this.state.desc || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount.'
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        desc: this.state.desc,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onFormSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}

        <input
          type="text"
          className="text-input"
          value={this.state.desc}
          placeholder="Description"
          onChange={this.onDescChange}
        />

        <input
          type="text"
          className="text-input"
          value={this.state.amount}
          placeholder="Amount"
          onChange={this.onAmountChange}
        />

        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.focused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          displayFormat="D/M/YYYY"
        />

        <textarea
          className="text-area"
          value={this.state.note}
          onChange={this.onNoteChange}
          placeholder="Add a note for this expense (optional)"
        />

        <div>
          <button className="button">Save</button>
        </div>
      </form>
    );
  }
}

ExpenseForm.propTypes = {
  desc: PropTypes.string,
  amount: PropTypes.number,
  note: PropTypes.string,
  onSubmit: PropTypes.func,
  createdAt: PropTypes.number
};

ExpenseForm.defaultProps = undefined;
