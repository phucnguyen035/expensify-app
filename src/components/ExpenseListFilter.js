import React, { Fragment, Component } from 'react';
import { connect } from 'react-redux';
import {
  setTextFilters,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

const mapStateToProps = state => ({
  filters: state.filters
});
const mapDispatchtoProps = dispatch => ({
  setTextFilters: text => dispatch(setTextFilters(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
});

export class ExpenseListFilter extends Component {
  state = {
    focusedInput: null
  };

  // Sort function
  onSortChange = sortValue => {
    if (sortValue === 'date') {
      this.props.sortByDate();
    } else if (sortValue === 'amount') {
      this.props.sortByAmount();
    }
  };

  onTextChange = text => {
    this.props.setTextFilters(text);
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  focusedInput = focusedInput => {
    this.setState(() => ({ focusedInput }));
  };

  render() {
    return (
      <Fragment>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={e => this.onTextChange(e.target.value)}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={e => this.onSortChange(e.target.value)}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId="startDate_filters"
          endDate={this.props.filters.endDate}
          endDateId="endDate_filters"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focusedInput}
          onFocusChange={this.focusedInput}
          isOutsideRange={() => false}
          numberOfMonths={1}
          showClearDates={true}
          displayFormat="DD/MM/YYYY"
        />
      </Fragment>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchtoProps
)(ExpenseListFilter);
