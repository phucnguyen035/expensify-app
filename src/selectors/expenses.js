import moment from 'moment';
// Visible expenses
export default (expenses, { text, startDate, endDate, sortBy }) =>
  expenses
    .filter((expense) => {
      const startDateMatch = startDate ? startDate.isSameOrBefore(moment(expense.createdAt)) : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(moment(expense.createdAt)) : true;
      const textMatch = expense.desc.toLowerCase().includes(text.toLowerCase());

      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
      // Default return latest item
      return a.createdAt < b.createdAt ? 1 : -1;
    });
