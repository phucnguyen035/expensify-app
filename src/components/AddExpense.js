import React, { Component } from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

export class AddExpense extends Component {
  onSubmit = expense => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h1>This is from my add expense component</h1>

        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpense);
