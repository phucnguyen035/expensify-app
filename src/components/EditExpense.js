import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});
const mapDispatchToProps = dispatch => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: id => dispatch(removeExpense(id))
});

export class EditExpense extends Component {
  onSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onClick = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm {...this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpense);
