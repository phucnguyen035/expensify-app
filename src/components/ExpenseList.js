import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import getVisibleExpenses from '../selectors/expenses';
import { ExpenseListItem } from './ExpenseListItem';
import ExpenseListFilter from './ExpenseListFilter';

const mapStateToProps = state => ({
  expenses: getVisibleExpenses(state.expenses, state.filters)
});

export const ExpenseList = props => (
  <Fragment>
    <ExpenseListFilter />

    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      props.expenses.map(expense => <ExpenseListItem {...expense} key={expense.id} />)
    )}
  </Fragment>
);

ExpenseList.propTypes = {
  expenses: PropTypes.arrayOf(PropTypes.any)
};

ExpenseList.defaultProps = undefined;

export default connect(mapStateToProps)(ExpenseList);
