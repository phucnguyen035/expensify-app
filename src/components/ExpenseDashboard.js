import React from 'react';
import ExpenseList from './ExpenseList';
import ExpensesSummary from './ExpensesSummary';
import ExpenseListFilter from './ExpenseListFilter';

const ExpenseDashboard = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default ExpenseDashboard;
