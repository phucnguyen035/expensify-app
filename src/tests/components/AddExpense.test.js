import React from 'react';
import { shallow } from 'enzyme';
import { AddExpense } from '../../components/AddExpense';
import expenses from '../fixtures/expenses';

describe('Add expense page', () => {
  let startAddExpense;
  let history;
  let wrapper;

  beforeEach(() => {
    startAddExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpense startAddExpense={startAddExpense} history={history} />);
  });

  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);

    expect(history.push).toHaveBeenLastCalledWith('/dashboard');
    expect(startAddExpense).toHaveBeenLastCalledWith(expenses[1]);
  });
});
