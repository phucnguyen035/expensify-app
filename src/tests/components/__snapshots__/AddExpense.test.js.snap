// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add expense page should render correctly 1`] = `
<div>
  <h1>
    This is from my add expense component
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
